File: daedalos-tools/journal/lib/narrative.py
"""Journal narrative builder - synthesizes events into readable stories."""

from datetime import datetime, timedelta
from typing import List, Dict, Any, Optional
from collections import defaultdict

try:
    from .collector import Event, collect_all_events
except ImportError:
    from collector import Event, collect_all_events


def format_time(ts: float) -> str:
    """Format timestamp for display."""
    dt = datetime.fromtimestamp(ts)
    now = datetime.now()

    if dt.date() == now.date():
        return dt.strftime("%H:%M:%S")
    elif dt.date() == (now - timedelta(days=1)).date():
        return "Yesterday " + dt.strftime("%H:%M")
    else:
        return dt.strftime("%Y-%m-%d %H:%M")


def format_duration(seconds: float) -> str:
    """Format duration for display."""
    if seconds < 60:
        return f"{int(seconds)}s"
    elif seconds < 3600:
        return f"{int(seconds // 60)}m {int(seconds % 60)}s"
    else:
        hours = int(seconds // 3600)
        minutes = int((seconds % 3600) // 60)
        return f"{hours}h {minutes}m"


def format_relative_time(ts: float) -> str:
    """Format timestamp as relative time."""
    now = datetime.now()
    dt = datetime.fromtimestamp(ts)
    delta = now - dt

    if delta.total_seconds() < 60:
        return "just now"
    elif delta.total_seconds() < 3600:
        mins = int(delta.total_seconds() / 60)
        return f"{mins} minute{'s' if mins != 1 else ''} ago"
    elif delta.total_seconds() < 86400:
        hours = int(delta.total_seconds() / 3600)
        return f"{hours} hour{'s' if hours != 1 else ''} ago"
    else:
        days = delta.days
        return f"{days} day{'s' if days != 1 else ''} ago"


def build_narrative(
    events: List[Event],
    verbose: bool = False,
) -> str:
    """
    Build a human-readable narrative from events.

    Args:
        events: List of events (should already be sorted newest-first)
        verbose: Include more details

    Returns:
        Formatted narrative string
    """
    if not events:
        return "No events found in the specified time range."

    lines = []

    # Group events by time period
    now = datetime.now()
    today_events = []
    yesterday_events = []
    older_events = []

    for event in events:
        dt = datetime.fromtimestamp(event.timestamp)
        if dt.date() == now.date():
            today_events.append(event)
        elif dt.date() == (now - timedelta(days=1)).date():
            yesterday_events.append(event)
        else:
            older_events.append(event)

    # Build narrative sections
    if today_events:
        lines.append("## Today")
        lines.append("")
        lines.extend(_format_event_group(today_events, verbose))

    if yesterday_events:
        lines.append("")
        lines.append("## Yesterday")
        lines.append("")
        lines.extend(_format_event_group(yesterday_events, verbose))

    if older_events:
        lines.append("")
        lines.append("## Earlier")
        lines.append("")
        lines.extend(_format_event_group(older_events, verbose))

    return "\n".join(lines)


def _format_event_group(events: List[Event], verbose: bool) -> List[str]:
    """Format a group of events."""
    lines = []

    for event in events:
        time_str = format_time(event.timestamp)
        icon = _get_icon(event.source, event.event_type)

        if verbose:
            lines.append(f"{time_str}  {icon} [{event.source}] {event.summary}")
            if event.details:
                for key, value in event.details.items():
                    if key not in ("timestamp", "source", "event_type", "summary"):
                        lines.append(f"           {key}: {value}")
        else:
            lines.append(f"{time_str}  {icon} {event.summary}")

    return lines


def _get_icon(source: str, event_type: str) -> str:
    """Get an icon for an event type."""
    icons = {
        # Sources
        "gates": "🚧",
        "loop": "🔄",
        "agent": "🤖",
        "undo": "↩️",
        "mcp-hub": "🔌",

        # Event types
        "gate_check": "🚧",
        "loop_started": "▶️",
        "loop_completed": "✅",
        "loop_failed": "❌",
        "loop_stopped": "⏹️",
        "agent_spawned": "🤖",
        "agent_killed": "💀",
        "file_created": "📄",
        "file_modified": "✏️",
        "file_deleted": "🗑️",
        "checkpoint": "📌",
    }

    return icons.get(event_type, icons.get(source, "•"))


def build_summary(
    events: List[Event],
) -> Dict[str, Any]:
    """
    Build a summary of events.

    Returns:
        Dictionary with summary statistics
    """
    summary = {
        "total_events": len(events),
        "time_range": {
            "start": None,
            "end": None,
        },
        "by_source": defaultdict(int),
        "by_type": defaultdict(int),
        "highlights": [],
    }

    if not events:
        return summary

    # Calculate stats
    timestamps = [e.timestamp for e in events]
    summary["time_range"]["start"] = format_time(min(timestamps))
    summary["time_range"]["end"] = format_time(max(timestamps))

    for event in events:
        summary["by_source"][event.source] += 1
        summary["by_type"][event.event_type] += 1

    # Identify highlights
    highlights = []

    # Check for loops
    loop_events = [e for e in events if e.source == "loop"]
    if loop_events:
        completed = len([e for e in loop_events if e.event_type == "loop_completed"])
        failed = len([e for e in loop_events if e.event_type == "loop_failed"])
        if completed:
            highlights.append(f"{completed} loop(s) completed")
        if failed:
            highlights.append(f"{failed} loop(s) failed")

    # Check for gate denials
    gate_events = [e for e in events if e.source == "gates"]
    denied = len([e for e in gate_events if "denied" in e.summary.lower()])
    if denied:
        highlights.append(f"{denied} gate check(s) denied")

    # Check for agents
    agent_events = [e for e in events if e.source == "agent"]
    if agent_events:
        highlights.append(f"{len(agent_events)} agent event(s)")

    # Check for file changes
    file_events = [e for e in events if e.source == "undo"]
    if file_events:
        checkpoints = len([e for e in file_events if e.event_type == "checkpoint"])
        changes = len(file_events) - checkpoints
        if changes:
            highlights.append(f"{changes} file change(s)")
        if checkpoints:
            highlights.append(f"{checkpoints} checkpoint(s)")

    summary["highlights"] = highlights
    summary["by_source"] = dict(summary["by_source"])
    summary["by_type"] = dict(summary["by_type"])

    return summary


def what_happened(
    hours: float = 1,
    sources: Optional[List[str]] = None,
    verbose: bool = False,
) -> str:
    """
    Answer "what happened?" for a time period.

    Args:
        hours: How many hours to look back
        sources: Filter by sources
        verbose: Include more details

    Returns:
        Human-readable narrative
    """
    events = collect_all_events(hours=hours, sources=sources)

    if not events:
        period = f"the last {hours} hour{'s' if hours != 1 else ''}"
        return f"Nothing recorded in {period}."

    summary = build_summary(events)

    lines = []

    # Header with summary
    lines.append(f"## What happened in the last {hours} hour{'s' if hours != 1 else ''}?")
    lines.append("")

    if summary["highlights"]:
        lines.append("**Highlights:** " + " • ".join(summary["highlights"]))
        lines.append("")

    lines.append(f"**{summary['total_events']} events** from {summary['time_range']['start']} to {summary['time_range']['end']}")
    lines.append("")

    # Activity by source
    if summary["by_source"]:
        sources_str = ", ".join(f"{k}: {v}" for k, v in sorted(summary["by_source"].items(), key=lambda x: -x[1]))
        lines.append(f"By source: {sources_str}")
        lines.append("")

    # Detailed narrative
    lines.append("---")
    lines.append("")
    lines.append(build_narrative(events, verbose=verbose))

    return "\n".join(lines)

File: crates/journal/src/narrative.rs
//! Narrative builder - synthesizes events into readable stories
//!
//! "What happened?" is the most important debugging question.
//! Raw events are overwhelming. A narrative is actionable.

use crate::db::Event;
use chrono::{Duration, Local, TimeZone, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Format timestamp for display
pub fn format_time(ts: f64) -> String {
    let dt = Utc.timestamp_opt(ts as i64, 0)
        .single()
        .unwrap_or_else(Utc::now);
    let local = dt.with_timezone(&Local);
    let now = Local::now();

    if local.date_naive() == now.date_naive() {
        local.format("%H:%M:%S").to_string()
    } else if local.date_naive() == (now - Duration::days(1)).date_naive() {
        format!("Yesterday {}", local.format("%H:%M"))
    } else {
        local.format("%Y-%m-%d %H:%M").to_string()
    }
}

/// Format duration in human-readable form
#[allow(dead_code)]
pub fn format_duration(seconds: f64) -> String {
    if seconds < 60.0 {
        format!("{}s", seconds as i64)
    } else if seconds < 3600.0 {
        let mins = (seconds / 60.0) as i64;
        let secs = (seconds % 60.0) as i64;
        format!("{}m {}s", mins, secs)
    } else {
        let hours = (seconds / 3600.0) as i64;
        let mins = ((seconds % 3600.0) / 60.0) as i64;
        format!("{}h {}m", hours, mins)
    }
}

/// Format timestamp as relative time (e.g., "2m ago")
#[allow(dead_code)]
pub fn format_relative_time(ts: f64) -> String {
    let dt = Utc.timestamp_opt(ts as i64, 0)
        .single()
        .unwrap_or_else(Utc::now);
    let now = Utc::now();
    let diff = now.signed_duration_since(dt);

    if diff.num_seconds() < 60 {
        "just now".to_string()
    } else if diff.num_minutes() < 60 {
        let mins = diff.num_minutes();
        format!("{} minute{} ago", mins, if mins != 1 { "s" } else { "" })
    } else if diff.num_hours() < 24 {
        let hours = diff.num_hours();
        format!("{} hour{} ago", hours, if hours != 1 { "s" } else { "" })
    } else {
        let days = diff.num_days();
        format!("{} day{} ago", days, if days != 1 { "s" } else { "" })
    }
}

/// Get an icon for an event source or type
fn get_icon(source: &str, event_type: &str) -> &'static str {
    // Note: Using simple ASCII characters for better terminal compatibility
    // Can be replaced with emojis if desired
    match event_type {
        "gate_check" => "[GATE]",
        "loop_started" => "[START]",
        "loop_completed" => "[DONE]",
        "loop_failed" => "[FAIL]",
        "loop_stopped" => "[STOP]",
        "agent_spawned" | "spawn" => "[AGENT]",
        "agent_killed" | "kill" => "[KILL]",
        "file_created" | "file_create" => "[+FILE]",
        "file_modified" | "file_edit" => "[~FILE]",
        "file_deleted" | "file_delete" => "[-FILE]",
        "checkpoint" => "[CHKPT]",
        _ => match source {
            "gates" => "[GATE]",
            "loop" => "[LOOP]",
            "agent" => "[AGENT]",
            "undo" => "[UNDO]",
            "mcp-hub" => "[MCP]",
            "user" | "journal" => "[USER]",
            _ => "[*]",
        },
    }
}

/// Summary statistics for events
#[derive(Debug, Serialize, Deserialize)]
pub struct Summary {
    pub total_events: usize,
    pub time_range: TimeRange,
    pub by_source: HashMap<String, usize>,
    pub by_type: HashMap<String, usize>,
    pub highlights: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TimeRange {
    pub start: Option<String>,
    pub end: Option<String>,
}

/// Build a summary of events
pub fn build_summary(events: &[Event]) -> Summary {
    let mut summary = Summary {
        total_events: events.len(),
        time_range: TimeRange {
            start: None,
            end: None,
        },
        by_source: HashMap::new(),
        by_type: HashMap::new(),
        highlights: Vec::new(),
    };

    if events.is_empty() {
        return summary;
    }

    // Calculate time range
    let timestamps: Vec<f64> = events.iter().map(|e| e.timestamp).collect();
    let min_ts = timestamps.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_ts = timestamps.iter().cloned().fold(f64::NEG_INFINITY, f64::max);

    summary.time_range.start = Some(format_time(min_ts));
    summary.time_range.end = Some(format_time(max_ts));

    // Count by source and type
    for event in events {
        *summary.by_source.entry(event.source.clone()).or_insert(0) += 1;
        *summary.by_type.entry(event.event_type.clone()).or_insert(0) += 1;
    }

    // Generate highlights
    let mut highlights = Vec::new();

    // Check for loops
    let loop_events: Vec<_> = events.iter().filter(|e| e.source == "loop").collect();
    if !loop_events.is_empty() {
        let completed = loop_events.iter().filter(|e| e.event_type == "loop_completed").count();
        let failed = loop_events.iter().filter(|e| e.event_type == "loop_failed").count();
        if completed > 0 {
            highlights.push(format!("{} loop(s) completed", completed));
        }
        if failed > 0 {
            highlights.push(format!("{} loop(s) failed", failed));
        }
    }

    // Check for gate denials
    let gate_events: Vec<_> = events.iter().filter(|e| e.source == "gates").collect();
    let denied = gate_events.iter().filter(|e| e.summary.to_lowercase().contains("denied")).count();
    if denied > 0 {
        highlights.push(format!("{} gate check(s) denied", denied));
    }

    // Check for agents
    let agent_events: Vec<_> = events.iter().filter(|e| e.source == "agent").collect();
    if !agent_events.is_empty() {
        highlights.push(format!("{} agent event(s)", agent_events.len()));
    }

    // Check for file changes
    let file_events: Vec<_> = events.iter().filter(|e| e.source == "undo").collect();
    if !file_events.is_empty() {
        let checkpoints = file_events.iter().filter(|e| e.event_type == "checkpoint").count();
        let changes = file_events.len() - checkpoints;
        if changes > 0 {
            highlights.push(format!("{} file change(s)", changes));
        }
        if checkpoints > 0 {
            highlights.push(format!("{} checkpoint(s)", checkpoints));
        }
    }

    summary.highlights = highlights;
    summary
}

/// Build a human-readable narrative from events
pub fn build_narrative(events: &[Event], verbose: bool) -> String {
    if events.is_empty() {
        return "No events found in the specified time range.".to_string();
    }

    let mut lines = Vec::new();
    let now = Local::now();

    // Group events by time period
    let mut today_events = Vec::new();
    let mut yesterday_events = Vec::new();
    let mut older_events = Vec::new();

    for event in events {
        let dt = Utc.timestamp_opt(event.timestamp as i64, 0)
            .single()
            .unwrap_or_else(Utc::now)
            .with_timezone(&Local);

        if dt.date_naive() == now.date_naive() {
            today_events.push(event);
        } else if dt.date_naive() == (now - Duration::days(1)).date_naive() {
            yesterday_events.push(event);
        } else {
            older_events.push(event);
        }
    }

    // Build narrative sections
    if !today_events.is_empty() {
        lines.push("## Today".to_string());
        lines.push(String::new());
        lines.extend(format_event_group(&today_events, verbose));
    }

    if !yesterday_events.is_empty() {
        lines.push(String::new());
        lines.push("## Yesterday".to_string());
        lines.push(String::new());
        lines.extend(format_event_group(&yesterday_events, verbose));
    }

    if !older_events.is_empty() {
        lines.push(String::new());
        lines.push("## Earlier".to_string());
        lines.push(String::new());
        lines.extend(format_event_group(&older_events, verbose));
    }

    lines.join("\n")
}

/// Format a group of events
fn format_event_group(events: &[&Event], verbose: bool) -> Vec<String> {
    let mut lines = Vec::new();

    for event in events {
        let time_str = format_time(event.timestamp);
        let icon = get_icon(&event.source, &event.event_type);

        if verbose {
            lines.push(format!("{}  {} [{}] {}", time_str, icon, event.source, event.summary));
            for (key, value) in &event.details {
                if !matches!(key.as_str(), "timestamp" | "source" | "event_type" | "summary") {
                    let value_str = match value {
                        serde_json::Value::String(s) => s.clone(),
                        _ => value.to_string(),
                    };
                    lines.push(format!("           {}: {}", key, value_str));
                }
            }
        } else {
            lines.push(format!("{}  {} {}", time_str, icon, event.summary));
        }
    }

    lines
}

/// Generate the "what happened" narrative
pub fn what_happened(events: &[Event], hours: f64, verbose: bool) -> String {
    if events.is_empty() {
        let period = if hours == 1.0 {
            "the last hour".to_string()
        } else {
            format!("the last {} hours", hours as i64)
        };
        return format!("Nothing recorded in {}.", period);
    }

    let summary = build_summary(events);
    let mut lines = Vec::new();

    // Header with summary
    let period = if hours == 1.0 {
        "the last hour".to_string()
    } else {
        format!("the last {} hours", hours as i64)
    };
    lines.push(format!("## What happened in {}?", period));
    lines.push(String::new());

    if !summary.highlights.is_empty() {
        lines.push(format!("**Highlights:** {}", summary.highlights.join(" | ")));
        lines.push(String::new());
    }

    let time_range = match (&summary.time_range.start, &summary.time_range.end) {
        (Some(start), Some(end)) => format!("{} to {}", start, end),
        _ => "unknown".to_string(),
    };
    lines.push(format!("**{} events** from {}", summary.total_events, time_range));
    lines.push(String::new());

    // Activity by source
    if !summary.by_source.is_empty() {
        let mut sources: Vec<_> = summary.by_source.iter().collect();
        sources.sort_by(|a, b| b.1.cmp(a.1));
        let sources_str: Vec<_> = sources.iter().map(|(k, v)| format!("{}: {}", k, v)).collect();
        lines.push(format!("By source: {}", sources_str.join(", ")));
        lines.push(String::new());
    }

    // Detailed narrative
    lines.push("---".to_string());
    lines.push(String::new());
    lines.push(build_narrative(events, verbose));

    lines.join("\n")
}

/// Format summary for display (non-JSON)
pub fn format_summary(summary: &Summary) -> String {
    let mut lines = Vec::new();

    lines.push(format!("Total events: {}", summary.total_events));

    let time_range = match (&summary.time_range.start, &summary.time_range.end) {
        (Some(start), Some(end)) => format!("{} - {}", start, end),
        _ => "N/A".to_string(),
    };
    lines.push(format!("Time range: {}", time_range));
    lines.push(String::new());

    if !summary.highlights.is_empty() {
        lines.push("Highlights:".to_string());
        for h in &summary.highlights {
            lines.push(format!("  * {}", h));
        }
        lines.push(String::new());
    }

    if !summary.by_source.is_empty() {
        lines.push("By source:".to_string());
        let mut sources: Vec<_> = summary.by_source.iter().collect();
        sources.sort_by(|a, b| b.1.cmp(a.1));
        for (source, count) in sources {
            lines.push(format!("  {}: {}", source, count));
        }
        lines.push(String::new());
    }

    if !summary.by_type.is_empty() {
        lines.push("By event type:".to_string());
        let mut types: Vec<_> = summary.by_type.iter().collect();
        types.sort_by(|a, b| b.1.cmp(a.1));
        for (etype, count) in types {
            lines.push(format!("  {}: {}", etype, count));
        }
    }

    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_time() {
        let now = Utc::now().timestamp() as f64;
        let formatted = format_time(now);
        assert!(formatted.contains(':'));
    }

    #[test]
    fn test_format_duration() {
        assert_eq!(format_duration(30.0), "30s");
        assert_eq!(format_duration(90.0), "1m 30s");
        assert_eq!(format_duration(3700.0), "1h 1m");
    }

    #[test]
    fn test_format_relative_time() {
        let now = Utc::now().timestamp() as f64;
        assert_eq!(format_relative_time(now), "just now");
    }

    #[test]
    fn test_build_summary_empty() {
        let summary = build_summary(&[]);
        assert_eq!(summary.total_events, 0);
        assert!(summary.highlights.is_empty());
    }

    #[test]
    fn test_build_narrative_empty() {
        let narrative = build_narrative(&[], false);
        assert!(narrative.contains("No events found"));
    }
}
